File: Pandacap.LowLevel/Data/TheirContent/ATProto/BlueskyPostFavorite.fs
namespace Pandacap.Data

open System
open System.ComponentModel.DataAnnotations
open Pandacap.PlatformBadges

type BlueskyPostFavoriteUser() =
    member val DID = "" with get, set
    member val Handle = "" with get, set

type BlueskyPostFavoriteImage() =
    member val CID = "" with get, set
    member val Alt = "" with get, set

type BlueskyPostFavorite() =
    [<Key>]
    member val Id = Guid.Empty with get, set

    member val CID = "" with get, set
    member val RecordKey = "" with get, set
    member val CreatedBy = new BlueskyPostFavoriteUser() with get, set
    member val CreatedAt = DateTimeOffset.MinValue with get, set
    member val FavoritedAt = DateTimeOffset.MinValue with get, set
    member val HiddenAt = nullDateTimeOffset with get, set
    member val Text = "" with get, set
    member val Images = new ResizeArray<BlueskyPostFavoriteImage>() with get, set

    interface IFavorite with
        member this.HiddenAt
            with get () = this.HiddenAt
             and set value = this.HiddenAt <- value

        member this.FavoritedAt = this.FavoritedAt

    interface IPost with
        member _.Platform = Bluesky
        member this.Url = $"https://bsky.app/profile/{this.CreatedBy.DID}/post/{this.RecordKey}"
        member this.DisplayTitle = TitleGenerator.FromBody(this.Text)
        member this.Id = $"{this.Id}"
        member this.InternalUrl = $"/ATProto/ViewBlueskyPost?did={this.CreatedBy.DID}&rkey={this.RecordKey}"
        member this.ExternalUrl = $"https://bsky.app/profile/{this.CreatedBy.DID}/post/{this.RecordKey}"
        member this.ProfileUrl = $"https://bsky.app/profile/{this.CreatedBy.DID}"
        member this.Thumbnails = [
            for image in this.Images do {
                new IPostThumbnail with
                    member _.AltText = image.Alt
                    member _.Url = $"/ATProto/GetBlob?did={this.CreatedBy.DID}&cid={image.CID}"
            }
        ]
        member this.PostedAt = this.CreatedAt
        member _.Usericon = null
        member this.Username = this.CreatedBy.Handle

File: Pandacap.Database/BlueskyPostFavorite.cs
using Pandacap.UI.Badges;
using Pandacap.UI.Elements;

namespace Pandacap.Database
{
    public class BlueskyPostFavorite : IFavorite
    {
        public Guid Id { get; set; } = Guid.Empty;

        public string CID { get; set; } = "";

        public string RecordKey { get; set; } = "";

        public class User
        {
            public string? PDS { get; set; }
            public string DID { get; set; } = "";
            public string Handle { get; set; } = "";
        }

        public User CreatedBy { get; set; } = new();

        public DateTimeOffset CreatedAt { get; set; }

        public DateTimeOffset FavoritedAt { get; set; }

        public DateTimeOffset? HiddenAt { get; set; }

        public string Text { get; set; } = "";

        public class Image
        {
            public string CID { get; set; } = "";
            public string Alt { get; set; } = "";
        }

        public List<Image> Images { get; set; } = [];

        Badge IPost.Badge => Badges.ATProto.WithText(CreatedBy.PDS ?? "public.api.bsky.app");

        string IPost.DisplayTitle => Text;

        string IPost.Id => $"{Id}";

        string IPost.InternalUrl => $"/ATProto/ViewBlueskyPost?did={CreatedBy.DID}&rkey={RecordKey}";

        string IPost.ExternalUrl => $"https://bsky.app/profile/{CreatedBy.DID}/post/{RecordKey}";

        DateTimeOffset IPost.PostedAt => CreatedAt;

        string? IPost.ProfileUrl => $"https://bsky.app/profile/{CreatedBy.DID}";

        IEnumerable<IPostThumbnail> IPost.Thumbnails => Images.Select(image => new PostThumbnailAdapter(CreatedBy, image));

        string? IPost.Username => CreatedBy.Handle;

        string? IPost.Usericon => null;

        private class PostThumbnailAdapter(User user, Image image) : IPostThumbnail
        {
            string IPostThumbnail.Url => $"/ATProto/GetBlob?did={user.DID}&cid={image.CID}";

            string IPostThumbnail.AltText => image.Alt;
        }
    }
}
